### DISCUSSION TITLE
Add Feature Gates

### DISCUSSION THREAD

**Discussion by jonathon2nd** [rancher/rke2#3194] [General]
**Is your feature request related to a problem? Please describe.**

Unable to add feature gates in Rancher created RKE2 clusters, like RKE

**Describe the solution you'd like**

Some way to add feature gates through UI or API

**Describe alternatives you've considered**

Kinda need for RKE2

**Additional context**

I have looked though https://docs.rke2.io/ and have found nothing.

https://rancher-users.slack.com/archives/C01PHNP149L/p1659037158151269?thread_ts=1659005780.732539&cid=C01PHNP149L

>I am also attempting to figure out how to enable feature gate on RKE2.
I have looked at
https://rancher.com/docs/rancher/v2.6/en/cluster-admin/editing-clusters/rke2-config-reference/
which leads to
https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/
--feature-gates <A list of 'key=true/false' pairs>
...
(DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See [kubelet-config-file](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/) for more information.)
I have tried a few ways to add, but all the ways I would add to rke do not seem to be there. I used to be able to add via API.
Wanting to add to test out [linstor-csi](https://github.com/piraeusdatastore/linstor-csi)
You will need to enable the following feature gates on both the kube-apiserver and all kubelets for this plugin to be operational: CSINodeInfo=true, CSIDriverRegistry=true,VolumeSnapshotDataSource=true. Please ensure that your version of Kubernetes is recent enough to enable these gates.
Other Doc: https://linbit.com/blog/linstor-csi-plugin-for-kubernetes/
Any know how to enable feature flags on RKE2 though Rancher?

**Comment by brandond** [comment]
https://docs.rke2.io/install/install_options/server_config/ 
```
   --kube-apiserver-arg value                    (flags) Customized flag for kube-apiserver process
   --kube-controller-manager-arg value           (flags) Customized flag for kube-controller-manager process
   --kube-scheduler-arg value                    (flags) Customized flag for kube-scheduler process
```

`--kube-apiserver-arg=FeatureGates=MyGate=true`

**Reply by yshashix** [reply]
Is it the same way to enable feature-gates on a recent version (v1.29) of RKE2??
Docs link you have posted doesn't work anymore.

**Comment by jonathon2nd** [comment]
So multiple would be `--kube-apiserver-arg=FeatureGates=MyGate1=true=MyGate2=true`?

**Comment by brandond** [comment]
A comma-delimited list, same as upstream: `--kube-apiserver-arg=FeatureGates=MyGate1=true,MyGate2=true`

Note that you will frequently need to add these feature-gates to multiple components - the apiserver, controller-manager, kubelet, and so on.

**Reply by hrvatskibogmars** [reply]
Can you provide an example of how to set this via config?

```
...
cni: calico

kube-apiserver-arg: "FeatureGates=CronJobTimeZone=true"
...

```

**Reply by hrvatskibogmars** [reply]
```
kube-apiserver-arg:
  - feature-gates=CronJobTimeZone=true
```
Found the solution. This works

**Comment by jonathon2nd** [comment]
Hmm, looking to enable IPVS for testing. Was able to modify kube-proxy on RKE no problem, not seeing where to do that for RKE2?

Edit:
I am meaning from Rancher UI. Before I took the value and put it into `Additional API Server Args`.
![image](https://user-images.githubusercontent.com/52681917/188895303-29913fe7-780b-41a6-b06f-9a0f67628eb9.png)

Trying by adding 
```
      kube-proxy-extra-env: 
        - proxy-mode=ipvs
      kube-proxy-extra-mount:
        - /lib/modules:/lib/modules:ro
```
Just below
```
    machineGlobalConfig:
      cni: calico
      disable:
      - rke2-ingress-nginx
```

**Reply by jonathon2nd** [reply]
hmm, it does look like this config is working, but only to a point. lib/modules are mounted on pods, but var is not being used right. Pods are still in iptables mode.
![image](https://user-images.githubusercontent.com/52681917/188899971-9ef8ca3d-5237-4f62-b078-8f027e1ad2fb.png)

**Reply by jonathon2nd** [reply]
yay found my mistake.
```
      kube-proxy-arg: 
        - proxy-mode=ipvs
      kube-proxy-extra-mount:
        - /lib/modules:/lib/modules:ro
```
Are the correct configs for enabling IPVS mode on RKE2. This can be set from cluster creation in Rancher UI. When creating new cluster and finished modifying the UI, click `edit as yaml` and inset the above based on context above.
